testsuite/shared/src/main/java/org/wildfly/test/stability/StabilityServerSetupSnapshotRestoreTasks.java

```diff
@@ -0,0 +1,179 @@
+        Assume.assumeTrue(
+                String.format("%s is not a supported stability level", desiredStability, supportedStabilityLevels),
+                supportedStabilityLevels.contains(desiredStability));
+
+        // Check the reload-enhanced operation exists in the current stability level
+        Assume.assumeTrue(
+                "The reload-enhanced operation is not registered at this stability level",
+                checkReloadEnhancedOperationIsAvailable(managementClient));
+
+        // Check the reload-enhanced operation exists in the stability level we want to load to so that
+        // we can reload back to the current one
+        Stability reloadOpStability = getReloadEnhancedOperationStabilityLevel(managementClient);
+        Assume.assumeTrue(desiredStability.enables(reloadOpStability));
+
+
+
+        originalStability = readCurrentStability(managementClient.getControllerClient());
+
+        // Take a snapshot, indicating that when reloading we want to go back to the original stability
+        snapshot = ServerSnapshot.takeSnapshot(managementClient, originalStability);
+
+        // All good, let's do it!
+        reloadToDesiredStability(managementClient.getControllerClient(), desiredStability);
+
+        // Do any additional setup from the sub-classes
+        doSetup(managementClient);
+    }
+
+    protected void doSetup(ManagementClient managementClient) throws Exception {
+
+    }
+
+    @Override
+    public void tearDown(ManagementClient managementClient) throws Exception {
+        snapshot.close();
+    }
+
+    private boolean checkReloadEnhancedOperationIsAvailable(ManagementClient managementClient) throws Exception {
+        ModelNode op = Util.createOperation(READ_OPERATION_NAMES_OPERATION, PathAddress.EMPTY_ADDRESS);
+        ModelNode result = ManagementOperations.executeOperation(managementClient.getControllerClient(), op);
+        for (ModelNode name : result.asList()) {
+            if (name.asString().equals(RELOAD_ENHANCED)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Stability getReloadEnhancedOperationStabilityLevel(ManagementClient managementClient) throws Exception {
+        ModelNode op = Util.createOperation(READ_OPERATION_DESCRIPTION_OPERATION, PathAddress.EMPTY_ADDRESS);
+        op.get(NAME).set(RELOAD_ENHANCED);
+
+        ModelNode result = ManagementOperations.executeOperation(managementClient.getControllerClient(), op);
+        String stability = result.get(STABILITY).asString();
+        return Stability.fromString(stability);
+
+    }
+    private Set<Stability> getSupportedStabilityLevels() {
+        // TODO WFCORE-6731 - see https://github.com/wildfly/wildfly-core/pull/5895#discussion_r1520489808
+        // This information will be available in a management operation, For now just return a hardcoded set
+        return EnumSet.allOf(Stability.class);
```

Comment:
I understand this is not going to block this PR. We could address it in a follow up.

WFCORE-6731 is in place and I think ready to be merged https://github.com/wildfly/wildfly-core/pull/5908, @pferraro FYI